base/sysinfo.jl

```diff
@@ -321,4 +323,88 @@ windows_version
 
 const WINDOWS_VISTA_VER = v"6.0"
 
+"""
+    Sys.isexecutable(path::String)
+
+Returns `true` if the given `path` has executable permissions.
+"""
+function isexecutable(path::String)
+    @static if iswindows()
+        return isfile(path)
+    else
+        # We use `access()` and `X_OK` to determine if a given path is
+        # executable by the current user.  `X_OK` comes from `unistd.h`.
+        X_OK = 0x01
+        ccall(:access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
+    end
+end
+isexecutable(path::AbstractString) = isexecutable(String(path))
+
+"""
+    Sys.which(program_name::String)
+
+Given a program name, searches the current `PATH` to find the first binary with
+the proper executable permissions that can be run, and returns the absolute
+path. Raises `ArgumentError` if no such program is available.  If a relative or
+absolute path is passed in for `program_name`, that exact path is tested for
+executable permissions only, no searching of `PATH` is performed.
+"""
+function which(program_name::String)
+    # If prog has a slash, we know the user wants to determine whether the given
+    # file exists and is executable, and to not search the `PATH`.  Note that
+    # Windows can have either `\\` or `/` in its paths:
+    dirseps = @static iswindows() ? ('/', '\\') : ('/', )
+
+    if any(occursin.(dirseps, program_name))
+        # If it does exist, check that it's executable or fail out
+        if !isexecutable(program_name)
+            throw(ArgumentError("$program_name is not executable"))
+        end
+
+        # If it all checks out, return the abspath
+        return abspath(program_name)
```

Comment:
The Window's rules aren't suppose to short-circuit like this – the `program_names` rules still apply in this case